### PR TITLE
New usage reporting

### PR DESCRIPTION
**Pull request by attiasas**
- [x] All [tests](https://github.com/jfrog/frogbot#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
---
Depends on:
https://github.com/jfrog/jfrog-client-go/pull/807
https://github.com/jfrog/jfrog-cli-core/pull/908

Reporting to the new usage report APIs for:
* Xray (from version 3.80)
* Ecosystem

You will only see now a debug log that we are sending Usage:
image
There is no response log for success.

In case an error occur you will see:
image

### CODE REVIEW

File: commands/utils/utils_test.go [hunk 1]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+	usageGroup.Wait()
+	assert.NoError(t, <-channel)
+
+	usageGroup.Add(1)
+	channel = make(chan error)
+	go func() {
+		channel <- ReportUsage("", &config.ServerDetails{ArtifactoryUrl: "http://httpbin.org/status/404"}, &usageGroup)
+	}()
+	usageGroup.Wait()
+	assert.Error(t, <-channel)
```
Comment (yahavi) [inline commands/utils/utils_test.go:82]: Let's use assert.ErrorContains to ensure the content of the error

File: commands/utils/utils_test.go [hunk 2]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+	go func() {
+		channel <- ReportUsage("", &config.ServerDetails{}, &usageGroup)
+	}()
+	usageGroup.Wait()
+	assert.NoError(t, <-channel)
+
+	usageGroup.Add(1)
+	channel = make(chan error)
+	go func() {
+		channel <- ReportUsage("", &config.ServerDetails{ArtifactoryUrl: "http://httpbin.org/status/404"}, &usageGroup)
```
Comment (yahavi) [inline commands/utils/utils_test.go:79]: httpbin is not stable. We may get flaky tests.
Let's create a new usage handler that returns 404.

File: commands/utils/utils_test.go [hunk 3]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+				},
+			},
+		},
+		Repository{
+			Params: Params{
+				Git: Git{
+					RepoName: "repository2",
+				},
+			},
+		},
```
Comment (yahavi) [inline commands/utils/utils_test.go:100]:
```suggestion
		Repository{Params: Params{Git: Git{RepoName: "repository1"}}},
		Repository{Params: Params{Git: Git{RepoName: "repository2"}}},
```

File: commands/utils/utils_test.go [hunk 4]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+		channel <- ReportUsageToXray("", &config.ServerDetails{XrayUrl: "http://httpbin.org/status/404"}, getDummyRepo(), &usageGroup)
+	}()
+	usageGroup.Wait()
 	assert.Error(t, <-channel)
 }
 
-// Create HTTP handler to mock an Artifactory server suitable for report usage requests
-func createUsageHandler(t *testing.T, commandName string) http.HandlerFunc {
+// Create HTTP handler to mock an Artifactory/Xray/Ecosystem server suitable for report usage requests
+func createUsageHandler(t *testing.T, commandName, clientId, clientId2 string) http.HandlerFunc {
```
Comment (yahavi) [inline commands/utils/utils_test.go:172]: This method became too long and not readable.
Let's create a usage handler for each server: createArtifactoryUsageHandler and createXrayUsageHandler.

File: commands/utils/utils_test.go [hunk 5]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+	go func() {
+		channel <- ReportUsageToXray("", &config.ServerDetails{XrayUrl: "http://httpbin.org/status/404"}, getDummyRepo(), &usageGroup)
+	}()
+	usageGroup.Wait()
 	assert.Error(t, <-channel)
 }
 
-// Create HTTP handler to mock an Artifactory server suitable for report usage requests
-func createUsageHandler(t *testing.T, commandName string) http.HandlerFunc {
+// Create HTTP handler to mock an Artifactory/Xray/Ecosystem server suitable for report usage requests
```
Comment (yahavi) [inline commands/utils/utils_test.go:171]: I couldn't find that it used for the Ecosystem server
```suggestion
// Create HTTP handler to mock an Artifactory/Xray server suitable for report usage requests
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: commands/utils/utils.go [hunk 6]
```diff
@@ -148,27 +152,117 @@ func Chdir(dir string) (cbk func() error, err error) {
...
+	if err != nil || len(reports) == 0 {
+		// Nothing to report
+		return
+	}
+	log.Debug(usage.ReportUsagePrefix + "Sending info to Ecosystem...")
+	err = errors.Join(err, xrayusage.SendEcosystemUsageReports(reports...))
+	return
+}
+
+func CreateEcosystemReports(commandName string, serverDetails *config.ServerDetails, repositories RepoAggregator) (reports []xrayusage.ReportEcosystemUsageData, err error) {
```
Comment (eyalbe4) [inline commands/utils/utils.go:242]: Can probably be made a private function.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I suggest encapsulating all the usage reporting code into a new UsageReporter struct, that can be initialized and executed. It also best to have this UsageReporter as part of jfrog-cli-core, so that we're able to utilize it in the future for JFrog CLI command as well, and not just for Frogbot.

File: commands/utils/utils.go [hunk 7]
```diff
@@ -148,27 +152,117 @@ func Chdir(dir string) (cbk func() error, err error) {
...
 }
 
-func ReportUsage(commandName string, serverDetails *config.ServerDetails, usageReportSent chan<- error) {
-	var err error
+func ReportUsageOnCommand(commandName string, serverDetails *config.ServerDetails, repositories RepoAggregator, usageGroup *sync.WaitGroup) {
+	(*usageGroup).Add(3)
+	// Ecosystem
+	go func() {
+		if e := ReportUsageToEcosystem(commandName, serverDetails, repositories, usageGroup); e != nil {
+			log.Debug("Ecosystem "+usage.ReportUsagePrefix, e.Error())
```
Comment (sverdlov93) [inline commands/utils/utils.go:160]: won't cause a double log?
on client and here?

File: commands/utils/utils_test.go [hunk 8]
```diff
@@ -46,41 +49,168 @@ func TestChdirErr(t *testing.T) {
...
+func createUsageHandler(t *testing.T, commandName, clientId, clientId2 string) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		// Artifactory version
 		if r.RequestURI == "/api/system/version" {
```
Comment (sverdlov93) [inline commands/utils/utils_test.go:175]: can be converted to switch

**Pull request merged**